secretsmanager/src/SecretRetriever.cpp

```diff
@@ -0,0 +1,91 @@
+// Concord
+//
+// Copyright (c) 2021 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License"). You may not use
+// this product except in compliance with the Apache 2.0 License.
+//
+// This product may include a number of subcomponents with separate copyright notices and license
+// terms. Your use of these subcomponents is subject to the terms and conditions of the
+// subcomponent's license, as noted in the LICENSE file.
+
+#include "SecretRetriever.hpp"
+#include <iostream>
+#include <iomanip>
+#include <fstream>
+#include <exception>
+#include <optional>
+#include <regex>
+#include <nlohmann/json.hpp>
+#include "httplib.h"
+
+using json = nlohmann::json;
+using namespace std::string_literals;
+
+namespace SecretRetriever {
+
+SecretData ParseJson(json j) {
+  return SecretData{j["password"].get<std::string>(),
+                    j["message_digest_algorithm"].get<std::string>(),
+                    j["encryption_algorithm"].get<std::string>(),
+                    (uint32_t)std::stoi(j["key_length"].get<std::string>())};
+}
+
+std::optional<SecretData> FileSecretRetrieve(const std::string &secret_url) {
+  std::regex url_regex(R"(^file://(.*))", std::regex::extended);
+  std::smatch url_match;
+  if (std::regex_match(secret_url, url_match, url_regex)) {
+    std::ifstream jsonFile(url_match[1], std::ifstream::in);
+    if (jsonFile.fail()) throw std::runtime_error("Secret retrieve failed to open "s + secret_url);
+    json j;
+    jsonFile >> j;
+    return ParseJson(j);
+  } else {
+    return std::nullopt;
+  }
+}
+
+std::optional<SecretData> RemoteSecretRetrieve(const std::string &secret_url) {
+  std::regex url_regex(R"(^(https|http)://([a-zA-Z0-9\.-]+)(:([0-9]+))?(.*))", std::regex::extended);
+  std::smatch url_match;
+  if (std::regex_match(secret_url, url_match, url_regex)) {
+    std::shared_ptr<httplib::Response> httpres;
+    const std::string host = url_match[2];
+    const std::string port = url_match[4];
+    const std::string path = url_match[5];
+
+    if ("http"s.compare(url_match[1]) == 0) {
+      int p = 80;
+      if (!port.empty()) p = stoi(port);
+      httplib::Client cli(host, p);
+      httpres = cli.Get(path.c_str());
+    } else {
+      int p = 443;
+      if (!port.empty()) p = stoi(port);
+      httplib::SSLClient cli(host, p);
+      cli.enable_server_certificate_verification(false);
```

Comment:
Why wouldn't we verify the server cert? Doesn't this open us up to a MIIM attack?

Is this because the server doesn't use PKI and we don't have a way to load the server public key locally first?

Reply:
I've added my comments below.

> @tdimitrov, @KTeofanidis thanks for the PR!
> 
> As far as I understand, the secret data decryption process includes three main steps:
> 
> 1. Secret key retrieval using TPM (SAAS team);
> 2. Encrypted data retrieval (local or remote file specified by URL) - `FileSecretRetrieve` and `RemoteSecretRetrieve`;

FileSecretRetrieve will be used only for tests. It's not intended for production use. We do can (and will) use the mock of the service for tests, but there was a hard requirement for a file interface too.

> 3. Encrypted data decryption using the key from  1. - `ISecretsManagerImpl & Sons`
> 
> There's one thing though I have to clarify.
> I'm not sure where the requirement on REST API came from, I haven't seen it in a design documents.
> Also I haven't seen any requirement about using JSON as an underlying format.

Both requirements are in the design document. SaaS will provide REST API running as a sidecar container and it will return JSON.

> In order to support all possible options, including mentioned above, as well as to avoid adding dependencies to concord-bft (httplib and nlohmann/json), I propose a full separation between secret data retrieval and decryption on an interface level:
> 
> 1. `ISecretManagerImpl` with its successors are already defined, although I'd rename them into `ISecretManager`, `SecretManagerAES` and `SecretManagerPlain` correspondingly;

This was my initial approach, but unfortunately the decryption process is not that simple. It combines key derivation and encryption, which are two separate algorithms. For this reason there is a single class which encapsulates different encryption and digest algorithms.

> 2. Define `ISecretRetriever` interface with an abstract method like `SecretData get(string URL);`
> 3. `FileSecretRetriever` will implement it in concord-bft (as per proposed PR);
> 4. `RemoteSecretRetriever` will implement it in concord (including httplib deps if decided, which are already there).

We can do this but I think it's not needed. At the moment the Retriever and the SecretsManagerEnc work as a pipeline. One initialises the retriever, gets the secrets parameter and passes them to the SecretsManagerEnc. So the way secrets are retrieved is fully decoupled from SecretsManagerEnc.

But even if we introduce interfaces - we'll have to provide implementations for REST/File endpoint, so we will need the dependencies somewhere in the project.

> 5. TPM interface should also be defined in concord-bft and implemented in concord.
The whole TPM iface is done by the SaaS team. Accessing TPM from virtual machine is hard, so they handle this from the outside.

> Secret data format is application dependent, i.e. is opaque to a SecretManager and known by its clients - `KeyManager`, etc.

Reply:
We have no way to obtain the Secrets API certificate in secure manner. HTTPS will not even be used in real deployments.